src/monodroid/jni/debug-app-helper.cc

```diff
@@ -196,30 +196,29 @@ runtime_exists (const char *dir, char*& libmonoso)
 	return false;
 }
 
-static char*
+static const char*
 get_libmonosgen_path ()
 {
 	char *libmonoso;
-	bool embedded_dso_mode_enabled = androidSystem.is_embedded_dso_mode_enabled ();
 
 	// Android 5 includes some restrictions on loading dynamic libraries via dlopen() from
 	// external storage locations so we need to file copy the shared object to an internal
 	// storage location before loading it.
 	copy_native_libraries_to_internal_location ();
 
-	if (!embedded_dso_mode_enabled) {
-		for (size_t i = 0; i < BasicAndroidSystem::MAX_OVERRIDES; ++i) {
-			if (runtime_exists (BasicAndroidSystem::override_dirs [i], libmonoso)) {
-				return libmonoso;
-			}
+	if (androidSystem.is_embedded_dso_mode_enabled ()) {
+		return SharedConstants::MONO_SGEN_SO;
```

Comment:
This may *appear* to work, but I don't understand *why* it works.  If we're in "embedded DSO mode" (feb9ea2a53,  6be4bdbcfd9f2fbe6267aceb934394bcdd0f13b4, bf657e815b200e57098725c0ad963d70e881ea69, …, 95ca1025f0, …?), then we need to *change* the path used for **dlopen**(3) to include the path to the `.apk`, a `!`, and then the path *within* the `.apk`.  See also: https://android.googlesource.com/platform/bionic/+/master/android-changes-for-ndk-developers.md

> Note that in API level 23 and above dlopen(3) will open a library from any zip file, not just your APK. Just give dlopen(3) a path of the form “my_zip_file.zip!/libs/libstuff.so”.

Thus, by my reading, *when* `androidSystem.is_embedded_dso_mode_enabled()`==true, we *can't* "just" `dlopen("libmonosgen-2.0.so")`, right…?

Unless we're implicitly depending on "Correct soname/path handling (Available in API level >= 23)" (same google doc link above):

> The dynamic linker now understands the difference between a library’s soname and its path…
> For example, `dlopen("/this/directory/does/not/exist/libc.so", RTLD_NOW)` would find `/system/lib/libc.so` because it’s already loaded.

*Because* we load `libmonosgen-2.0.so` very early in Java bootstrap, an "unpathed" `dlopen("libmonosgen-2.0.so")` would find the already loaded file, which is good.

…except that we currently support down to API-19, and with .NET 6 we'll support down to API-21, *both* of which are < API-23, which adds "Correct soname/path handling".

Thus, I *strongly suspect* that if you tried this codepath on an API-21 device/emulator, it would *not* work.

Reply:
@jonpryor embedded DSOs aren't available in API 21, they were added starting from API 23 - together with the correct search path handling for `dlopen`. So `androidSystem.is_embedded_dso_mode_enabled()` will always return `false` in API levels older than 23 and things will work as expected.

Reply:
ooh, good point.

Reply:
Actually, no, I don't get it. if embedded DSO mode is enabled *and* we're using `java_runtime_fastdev.jar`, in which case there is no `System.loadLibrary("monosgen-2.0")`, then *how* does `dlopen("libmonosgen-2.0.so")` *work*?

…and if we're using `java_runtime.jar`, then *we won't get here*, because this code is in `xamarin-debug-app-helper`, which isn't used when `BuildConfig.Debug`==false.

Apparently this works; we don't see `DebuggingTest.ApplicationRunsWithoutDebugger()` failing on CI for this PR.

I still don't understand *why* it works. :-(

Reply:
`dlopen` searches the equivalent of `LD_LIBRARY_PATH` which automatically includes the `myapp.apk!/libs/[ARCH]` location, so if you do `dlopen("somelib.so")` it will find it if it's there.